hack/tools/release/notes/print.go

```diff
@@ -0,0 +1,172 @@
+// for the release notes.
+type releaseNotesPrinter struct {
+	outputOrder            []string
+	isPreRelease           bool
+	printKubernetesSupport bool
+	printDeprecation       bool
+	fromTag                string
+	repo                   string
+}
+
+func newReleaseNotesPrinter(repo, fromTag string) *releaseNotesPrinter {
+	return &releaseNotesPrinter{
+		repo:        repo,
+		fromTag:     fromTag,
+		outputOrder: defaultOutputOrder,
+	}
+}
+
+// print outputs to stdout the release notes.
+func (p *releaseNotesPrinter) print(entries []notesEntry) {
+	merges := map[string][]string{
+		release.Features:      {},
+		release.Bugs:          {},
+		release.Documentation: {},
+		release.Warning:       {},
+		release.Other:         {},
+		release.Unknown:       {},
+	}
+
+	for _, entry := range entries {
+		if entry.section == release.Documentation {
+			merges[entry.section] = append(merges[entry.section], "#"+entry.prNumber)
+		} else {
+			merges[entry.section] = append(merges[entry.section], entry.title)
+		}
+	}
+
+	if p.isPreRelease {
+		fmt.Printf("🚨 This is a RELEASE CANDIDATE. Use it only for testing purposes. If you find any bugs, file an [issue](https://github.com/%s/issues/new).\n", p.repo)
+	}
+
+	if p.printKubernetesSupport {
+		fmt.Print(`## 👌 Kubernetes version support
+
+- Management Cluster: v1.**X**.x -> v1.**X**.x
+- Workload Cluster: v1.**X**.x -> v1.**X**.x
+
+[More information about version support can be found here](https://cluster-api.sigs.k8s.io/reference/versions.html)
+
+`)
+	}
+
+	fmt.Print(`## Highlights
+
+* REPLACE ME
+
+`)
+
+	if p.printDeprecation {
+		fmt.Print(`## Deprecation Warning
+
+REPLACE ME: A couple sentences describing the deprecation, including links to docs.
+
+* [GitHub issue #REPLACE ME](REPLACE ME)
+
+`)
+	}
+
+	fmt.Printf("## Changes since %s\n", p.fromTag)
+
+	fmt.Printf("## :chart_with_upwards_trend: Overview\n")
+	if count := len(entries); count == 1 {
+		fmt.Println("- 1 new commit merged")
+	} else if count > 1 {
+		fmt.Printf("- %d new commits merged\n", count)
+	}
+	if count := len(merges[release.Warning]); count == 1 {
+		fmt.Println("- 1 breaking change :warning:")
+	} else if count > 1 {
+		fmt.Printf("- %d breaking changes :warning:\n", count)
+	}
+	if count := len(merges[release.Features]); count == 1 {
+		fmt.Println("- 1 feature addition ✨")
+	} else if count > 1 {
+		fmt.Printf("- %d feature additions ✨\n", count)
+	}
+	if count := len(merges[release.Bugs]); count == 1 {
+		fmt.Println("- 1 bug fixed 🐛")
+	} else if count > 1 {
+		fmt.Printf("- %d bugs fixed 🐛\n", count)
+	}
+	fmt.Println()
+
+	for _, key := range p.outputOrder {
+		mergeslice := merges[key]
+		if len(mergeslice) == 0 {
+			continue
+		}
+
+		switch key {
+		case release.Documentation:
+			sort.Strings(mergeslice)
+			if len(mergeslice) == 1 {
+				fmt.Printf(
+					":book: Additionally, there has been 1 contribution to our documentation and book. (%s) \n\n",
+					mergeslice[0],
+				)
+			} else {
```

Comment:
Shouldn't this block also work if `len(mergeslice) == 1`? I could be misunderstanding how `strings.Join()` works in Go though

Reply:
The message changes
`there has been 1 contribution` vs `there has been 1 contributions`